src/lib.rs

```diff
@@ -23,161 +34,280 @@ use bevy::render::{
-    info!("Loaded debug lines plugin.");
+/// Move the DebugLinesMesh marker Component to the render context.
+fn extract_debug_lines(mut commands: Commands, query: Query<Entity, With<DebugLinesMesh>>) {
+    for entity in query.iter() {
+        commands.get_or_spawn(entity).insert(RenderDebugLinesMesh);
+    }
 }
 
-/// Shader data, passed to the GPU.
-#[derive(RenderResources, Default, ShaderDefs, TypeUuid)]
-#[uuid = "f093e7c5-634c-45f8-a2af-7fcd0245f259"]
-pub struct LineShader {
-    pub num_lines: u32, // max number of lines: see: MAX_LINES.
-    // I don't love having 2 buffers here.  It would be cleaner if we can do a custom line structure.
-    // We should also consider the memory imprint here.  We should maybe instead allow a predefined
-    // set of colors which would dramatically reduce that.
-    #[render_resources(buffer)]
-    pub points: Vec<Vec4>,
-    #[render_resources(buffer)]
-    pub colors: Vec<Vec4>,
-
-    #[render_resources(ignore)]
-    #[shader_def]
-    pub depth_test: bool,
+/// Marker Component for the [`Entity`] associated with the meshes rendered with the
+/// debuglines.wgsl shader.
+///
+/// Stores the index of the mesh for the logic of `ImmLinesStorage` and `RetLinesStorage`
+#[derive(Component)]
+enum DebugLinesMesh {
+    /// Meshes for duration=0.0 lines
+    Immediate(usize),
+    /// Meshes for duration≠0.0 lines
+    Retained(usize),
+}
+impl Default for DebugLinesMesh {
+    fn default() -> Self {
+        DebugLinesMesh::Immediate(0)
+    }
 }
 
-/// A single line, usually initialized by helper methods on `DebugLines` instead of directly.
-pub struct Line {
-    start: Vec3,
-    end: Vec3,
-    color: [Color; 2],
-    duration: f32,
+#[derive(Component)]
+struct RenderDebugLinesMesh;
+
+// NOTE: consider this: we could just hold a Handle<Mesh> to the DebugLinesMesh
+// and modify it in-place, so that there is no need to update the mesh every
+// frame on top of keeping track of all those buffers in `LinesStorage`.
+// However, I implemented that, and found out it was about 3 times slower in
+// the `bench.rs` example. Probably the 2 levels of indirection and the error
+// checking is what kills it.
+//
+// TODO: Use a u32 for colors, this may improve performance if decoding and
+// encoding the color is not more expensive than moving 4 values in memory 3
+// times
+/// The [`DebugLines`] storage for immediate mod lines.
+///
+/// This is `pub` because of the `SystemParam` macro on [`DebugLines`]. The end
+/// user **should absolutely not interact with this**.
+#[derive(Debug, Default)]
+#[doc(hidden)]
+pub struct ImmLinesStorage {
+    positions: Vec<[f32; 3]>,
+    colors: Vec<[f32; 4]>,
 }
+impl ImmLinesStorage {
+    fn add_at(&mut self, i: usize, start: Vec3, end: Vec3, start_color: Color, end_color: Color) {
+        self.colors[i] = start_color.into();
+        self.colors[i + 1] = end_color.into();
+        self.positions[i] = start.into();
+        self.positions[i + 1] = end.into();
+    }
```

Comment:
Yes it's not ideal. However it is necessary. I do this so that the `colors` and `positions` Vec have the exact same layout as the `Mesh` buffers, ie: each entry is the two vertices of the line. This is so copying the bits to the attribute buffers is a simple `memcpy`, see calls to `Mesh::attribute_mut` in the code and methods `fill_attributes`.

Reply:
This should be solved with #ced553bb3f23fa6